lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+  BundleType originalType = op.result().getType().cast<BundleType>();
+
+  // Create a new, flat bundle type for the new result
+  SmallVector<BundleType::BundleElement, 8> bundleElements;
+  for (auto element : originalType.getElements()) {
+    if (element.second.isa<BundleType>()) {
+      // Flatten nested bundle types the usual way.
+      SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+      flattenBundleTypes(element.second, element.first, false, fieldTypes);
+
+      for (auto field : fieldTypes) {
+        // Store the flat type for the new bundle type.
+        auto flatName = builder->getIdentifier(field.suffix);
+        auto flatType = getPortType(field);
+        auto newElement = BundleType::BundleElement(flatName, flatType);
+        bundleElements.push_back(newElement);
+      }
+    } else {
+      // Store the original type for the new bundle type.
+      bundleElements.push_back(element);
+    }
+  }
+
+  // Get the new bundle type and create a new instance.
+  auto newType = BundleType::get(bundleElements, &getContext());
+
+  auto newInstance = builder->create<InstanceOp>(
+      op.getLoc(), newType, op.moduleNameAttr(), op.nameAttr());
+
+  // Store the mapping from the old instance to the new one.
+  setInstanceLowering(op, newInstance);
+
+  removeOp(op);
+
+  return success();
+}
+
+// Lowering subfield operations has to deal with three different cases:
+//   1. the input value is from a module block argument
+//   2. the input value is from another subfield operation's result
+//   3. the input value is from an instance
+LogicalResult FIRRTLTypesLowering::visitExpr(SubfieldOp op) {
+  Value input = op.input();
+  Value result = op.result();
+  StringRef fieldname = op.fieldname();
+
+  Value rootBundle;
+  SmallString<16> suffix;
+  if (input.isa<BlockArgument>()) {
+    // When the input is from a block argument, map the result to the root
+    // bundle block argument, and begin creating a suffix string.
+    rootBundle = input;
+    suffix.push_back('_');
+    suffix += fieldname;
+  } else if (input.isa<OpResult>()) {
+    Operation *owner = input.cast<OpResult>().getOwner();
+    if (isa<SubfieldOp>(owner)) {
+      // When the input is from a subfield, look up the lowered info from the
+      // parent.
+      auto subfieldLowering = getSubfieldLowering(input);
+      assert(subfieldLowering.hasValue() && "no subfield lowering for input");
+
+      // Map the result to the root bundle block argument, and append the
+      // suffix to what was already mapped.
+      auto subfieldInfo = subfieldLowering.getValue();
+      rootBundle = subfieldInfo.first;
+      suffix.assign(subfieldInfo.second);
+      suffix.push_back('_');
+      suffix.append(fieldname);
+    } else if (isa<InstanceOp>(owner)) {
+      // When the input is from an instance, look up the new instance.
+      Optional<Value> newInstance = getInstanceLowering(input);
+      assert(newInstance.hasValue() && "no instance lowering for input");
```

Comment:
`getInstanceLowering` returns a non-optional Value, so you can remove the Optional and hasValue/getValue stuff.